File: 15_Boxing/javascript/boxing.js
// BOWLING
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(33) + "BOXING\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("BOXING OLYMPIC STYLE (3 ROUNDS -- 2 OUT OF 3 WINS)\n");
    j = 0;
    l = 0;
    print("\n");
    print("WHAT IS YOUR OPPONENT'S NAME");
    js = await input();
    print("INPUT YOUR MAN'S NAME");
    ls = await input();
    print("DIFFERENT PUNCHES ARE: (1) FULL SWING; (2) HOOK; (3) UPPERCUT; (4) JAB.\n");
    print("WHAT IS YOUR MANS BEST");
    b = parseInt(await input());
    print("WHAT IS HIS VULNERABILITY");
    d = parseInt(await input());
    do {
        b1 = Math.floor(4 * Math.random() + 1);
        d1 = Math.floor(4 * Math.random() + 1);
    } while (b1 == d1) ;
    print(js + "'S ADVANTAGE IS " + b1 + " AND VULNERABILITY IS SECRET.\n");
    print("\n");
    knocked = 0;
    for (r = 1; r <= 3; r++) {
        if (j >= 2)
            break;
        if (l >= 2)
            break;
        x = 0;
        y = 0;
        print("ROUND " + r + " BEGIN...\n");
        for (r1 = 1; r1 <= 7; r1++) {
            i = Math.floor(10 * Math.random() + 1);
            if (i <= 5) {
                print(ls + "'S PUNCH");
                p = parseInt(await input());
                if (p == b)
                    x += 2;
                if (p == 1) {
                    print(ls + " SWINGS AND ");
                    x3 = Math.floor(30 * Math.random() + 1);
                    if (d1 == 4 || x3 < 10) {
                        print("HE CONNECTS!\n");
                        if (x > 35) {
                            r = 3;
                            break;
                        }
                        x += 15;
                    } else {
                        print("HE MISSES \n");
                        if (x != 1)
                            print("\n\n");
                    }
                } else if (p == 2) {
                    print(ls + " GIVES THE HOOK... ");
                    h1 = Math.floor(2 * Math.random() + 1);
                    if (d1 == 2) {
                        x += 7;
                    } else if (h1 != 1) {
                        print("CONNECTS...\n");
                        x += 7;
                    } else {
                        print("BUT IT'S BLOCKED!!!!!!!!!!!!!\n");
                    }
                } else if (p == 3) {
                    print(ls + " TRIES AN UPPERCUT ");
                    d5 = Math.floor(100 * Math.random() + 1);
                    if (d1 == 3 || d5 < 51) {
                        print("AND HE CONNECTS!\n");
                        x += 4;
                    } else {
                        print("AND IT'S BLOCKED (LUCKY BLOCK!)\n");
                    }
                } else {
                    print(ls + " JABS AT " + js + "'S HEAD ");
                    c = Math.floor(8 * Math.random() + 1);
                    if (d1 == 4 || c >= 4) {
                        x += 3;
                    } else {
                        print("IT'S BLOCKED.\n");
                    }
                }
            } else {
                j7 = Math.random(4 * Math.random() + 1);
                if (j7 == b1)
                    y += 2;
                if (j7 == 1) {
                    print(js + " TAKES A FULL SWING AND");
                    r6 = Math.floor(60 * Math.random() + 1);
                    if (d == 1 || r6 < 30) {
                        print(" POW!!!!! HE HITS HIM RIGHT IN THE FACE!\n");
                        if (y > 35) {
                            knocked = 1;
                            r = 3;
                            break;
                        }
                        y += 15;
                    } else {
                        print(" IT'S BLOCKED!\n");
                    }
                } else if (j7 == 2 || j7 == 3) {
                    if (j7 == 2) {
                        print(js + " GETS " + ls + " IN THE JAW (OUCH!)\n");
                        y += 7;
                        print("....AND AGAIN!\n");
                        y += 5;
                        if (y > 35) {
                            knocked = 1;
                            r = 3;
                            break;
                        }
                        print("\n");
                        // From original, it goes over from handling 2 to handling 3
                    }
                    print(ls + " IS ATTACKED BY AN UPPERCUT (OH,OH)...\n");
                    q4 = Math.floor(200 * Math.random() + 1);
                    if (d == 3 || q4 <= 75) {
                        print("AND " + js + " CONNECTS...\n");
                        y += 8;
                    } else {
                        print(" BLOCKS AND HITS " + js + " WITH A HOOK.\n");
                        x += 5;
                    }
                } else {
                    print(js + " JABS AND ");
                    z4 = Math.floor(7 * Math.random() + 1);
                    if (d == 4)
                        y += 5;
                    else if (z4 > 4) {
                        print(" BLOOD SPILLS !!!\n");
                        y += 5;
                    } else {
                        print("IT'S BLOCKED!\n");
                    }
                }
            }
        }
        if (x > y) {
            print("\n");
            print(ls + " WINS ROUND " + r + "\n");
            l++;
        } else {
            print("\n");
            print(js + " WINS ROUND " + r + "\n");
            j++;
        }
    }
    if (j >= 2) {
        print(js + " WINS (NICE GOING, " + js + ").\n");
    } else if (l >= 2) {
        print(ls + " AMAZINGLY WINS!!\n");
    } else if (knocked) {
        print(ls + " IS KNOCKED COLD AND " + js + " IS THE WINNER AND CHAMP!\n");
    } else {
        print(js + " IS KNOCKED COLD AND " + ls + " IS THE WINNER AND CHAMP!\n");
    }
    print("\n");
    print("\n");
    print("AND NOW GOODBYE FROM THE OLYMPIC ARENA.\n");
    print("\n");
}

main();

File: 15_Boxing/perl/boxing.pl
#!/usr/bin/perl

# Boxing program in Perl
#   Required extensive restructuring to remove all of the GOTO's.
# Translated by Kevin Brannen (kbrannen)

use strict;
use warnings;

# globals
my $Opp_won = 0;    # num rounds opponent has won
my $You_won = 0;    # num rounds you have won
my $Opp_name = "";  # opponent name
my $Your_name = ""; # your name
my $Your_best = 0;  # your best punch
my $Your_worst = 0; # your worst punch
my $Opp_best;       # opponent best punch
my $Opp_worst;      # opponent worst punch
my $Opp_damage;     # opponent damage ?
my $Your_damage;    # your damage ?

sub get_punch
{
    my $prompt = shift;
    my $p;
    while (1)
    {
        print "$prompt: ";
        chomp($p = int(<>));
        last if ($p >= 1 && $p <= 4);
        print "DIFFERENT PUNCHES ARE: (1) FULL SWING; (2) HOOK; (3) UPPERCUT; (4) JAB.\n";
    }
    return $p;
}

print "\n";
print " " x 33, "BOXING\n";
print " " x 15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n\n\n";

print "BOXING OLYMPIC STYLE (3 ROUNDS -- 2 OUT OF 3 WINS)\n\n";
print "WHAT IS YOUR OPPONENT'S NAME: ";
chomp($Opp_name = <>);
print "INPUT YOUR MAN'S NAME: ";
chomp($Your_name = <>);
print "DIFFERENT PUNCHES ARE: (1) FULL SWING; (2) HOOK; (3) UPPERCUT; (4) JAB.\n";
$Your_best = get_punch("WHAT IS YOUR MANS BEST");
$Your_worst = get_punch("WHAT IS HIS VULNERABILITY");

do {
    $Opp_best = int(4*rand(1)+1);
    $Opp_worst = int(4*rand(1)+1);
} while ($Opp_best == $Opp_worst);
print "$Opp_name\'S ADVANTAGE IS $Opp_best AND VULNERABILITY IS SECRET.\n\n";

for my $R (1 .. 3) # rounds
{
    last if ($Opp_won >= 2 || $You_won >= 2);
    $Opp_damage = 0;
    $Your_damage = 0;
    print "ROUND $R BEGINS...\n";
    for my $R1 (1 .. 7) # 7 events per round?
    {
        if (int(10*rand(1)+1) <= 5)
        {
            my $your_punch = get_punch("$Your_name\'S PUNCH");
            $Opp_damage += 2 if ($your_punch == $Your_best);

            if    ($your_punch == 1) { punch1(); }
            elsif ($your_punch == 2) { punch2(); }
            elsif ($your_punch == 3) { punch3(); }
            else                     { punch4(); }
            next;
        }

        my $Opp_punch = int(4*rand(1)+1);
        $Your_damage += 2 if ($Opp_punch  == $Opp_best);
            
        if    ($Opp_punch == 1) { opp1(); }
        elsif ($Opp_punch == 2) { opp2(); }
        elsif ($Opp_punch == 3) { opp3(); }
        else                    { opp4(); }
    }

    if ($Opp_damage > $Your_damage)
    {
        print "\n$Your_name WINS ROUND $R\n\n";
        $You_won++;
    }
    else
    {
        print "\n$Opp_name WINS ROUND $R\n\n";
        $Opp_won++;
    }
}

if ($Opp_won >= 2)
{
    done("$Opp_name WINS (NICE GOING, $Opp_name).");
}

#else # if ($You_won >= 2)
done("$Your_name AMAZINGLY WINS!!");

###################################################

sub done
{
    my $msg = shift;
    print $msg;
    print "\n\nAND NOW GOODBYE FROM THE OLYMPIC ARENA.\n\n";
    exit(0);
}

sub punch1
{
    # $your_punch == 1, full swing
    print "$Your_name SWINGS AND ";
    if ($Opp_worst == 4 || int(30*rand(1)+1) < 10)
    {
        print "HE CONNECTS!\n";
        if ($Opp_damage > 35)
        {
            done("$Opp_name IS KNOCKED COLD AND $Your_name IS THE WINNER AND CHAMP! ");
        }
        $Opp_damage += 15;
    }
    else
    {
        print "HE MISSES\n";
        print "\n\n" if ($Opp_damage != 1);
    }
}

sub punch2
{
    # $your_punch == 2, hook
    print "$Your_name GIVES THE HOOK... ";
    if ($Opp_worst == 2)
    {
        $Opp_damage += 7;
        return;
    }
    if (int(2*rand(1)+1) == 1)
    {
        print "BUT IT'S BLOCKED!!!!!!!!!!!!!\n";
    }
    else
    {
        print "CONNECTS...\n";
        $Opp_damage += 7;
    }
}

sub punch3
{
    # $your_punch == 3, uppercut
    print "$Your_name TRIES AN UPPERCUT ";
    if ($Opp_worst == 3 || int(100*rand(1)+1) < 51)
    {
        print "AND HE CONNECTS!\n";
        $Opp_damage += 4;
    }
    else
    {
        print "AND IT'S BLOCKED (LUCKY BLOCK!)\n";
    }
}

sub punch4
{
    # $your_punch == 4, jab
    print "$Your_name JABS AT $Opp_name\'S HEAD ";
    if ($Opp_worst == 4 || (int(8*rand(1)+1)) >= 4)
    {
        $Opp_damage += 3;
        print "\n";
    }
    else
    {
        print "IT'S BLOCKED.\n";
    }
}

sub opp1
{
    # opp_punch == 1
    print "$Opp_name TAKES A FULL SWING AND ";
    if ($Your_worst == 1 || int(60*rand(1)+1) < 30)
    {
        print " POW!!!!! HE HITS HIM RIGHT IN THE FACE!\n";
        if ($Your_damage > 35)
        {
            done("$Your_name IS KNOCKED COLD AND $Opp_name IS THE WINNER AND CHAMP!");
        }
        $Your_damage += 15;
    }
    else
    {
        print " IT'S BLOCKED!\n";
    }
}

sub opp2
{
    # opp_punch == 2
    print "$Opp_name GETS $Your_name IN THE JAW (OUCH!)\n";
    $Your_damage += 7;
    print "....AND AGAIN!\n";
    $Your_damage += 5;
    if ($Your_damage > 35)
    {
        done("$Your_name IS KNOCKED COLD AND $Opp_name IS THE WINNER AND CHAMP!");
    }
    print "\n";
    # 2 continues into opp_punch == 3
    opp3();
}

sub opp3()
{
    # opp_punch == 3
    print "$Your_name IS ATTACKED BY AN UPPERCUT (OH,OH)...\n";
    if ($Your_worst != 3 && int(200*rand(1)+1) > 75)
    {
        print " BLOCKS AND HITS $Opp_name WITH A HOOK.\n";
        $Opp_damage += 5;
    }
    else
    {
        print "AND $Opp_name CONNECTS...\n";
        $Your_damage += 8;
    }
}

sub opp4
{
    # opp_punch == 4
    print "$Opp_name JABS AND ";
    if ($Your_worst == 4)
    {
        $Your_damage += 5;
    }
    elsif (int(7*rand(1)+1) > 4)
    {
        print " BLOOD SPILLS !!!\n";
        $Your_damage += 5;
    }
    else
    {
        print "IT'S BLOCKED!\n";
    }
}
